desktop/renderer/remote-pinentry.js

```diff
@@ -0,0 +1,67 @@
+// @flow
+import React, {Component} from 'react'
+import {connect} from 'react-redux'
+import {bindActionCreators} from 'redux'
+import RemoteComponent from './remote-component'
+import {registerPinentryListener, onCancel, onSubmit} from '../shared/actions/pinentry'
+import type {PinentryState} from '../shared/reducers/pinentry'
+import type {GUIEntryFeatures} from '../shared/constants/types/flow-types'
+
+type Props = {
+  registerPinentryListener: () => void,
+  onCancel: (sessionID: number) => void,
+  onSubmit: (sessionID: number, passphrase: string, features: GUIEntryFeatures) => void,
+  pinentryStates: {[key: string]: PinentryState}
+}
+
+class RemotePinentry extends Component<void, Props, void> {
+  componentWillMount () {
+    this.props.registerPinentryListener()
+  }
+
+  shouldComponentUpdate (nextProps, nextState) {
+    if (nextProps.pinentryStates !== this.props.pinentryStates) {
+      return true
+    }
+    return false
+  }
+
+  render () {
+    const {pinentryStates} = this.props
+
+    return (
+      <div>
+        {Object.keys(pinentryStates).filter(sid => !pinentryStates[sid].closed).map(pSessionID => {
+          const sid = parseInt(pSessionID, 10)
+          const onCancel = () => this.props.onCancel(sid)
+          const onSubmit = this.props.onSubmit.bind(null, sid)
```

Comment:
all this is me just splitting up and refactoring the code. i didn't want to change the logic of it that much as i didn't want to break a bunch of code. see master: https://github.com/keybase/client/blob/master/desktop/renderer/remote-manager.js#L103